### DISCUSSION TITLE
Make ORM update on instance consistent with add and delete

### DISCUSSION THREAD

**Discussion by januszm** [sqlalchemy/sqlalchemy#7483] [Ideas]
### Describe the use case

It's a bit strange that the ORM version of the `update` method is very different in use than how `insert` or `delete` is performed. When adding or removing a record, call `session.add(record)` or `session.delete(record)`. In the case of `update`, this is a completely different approach and, I must admit, not very intuitive. I think it would be easier for programmers to work with it if the update could also be performed on an already loaded record, i.e. `session.update (record) .values(**attributes_dict)`.
Unless I'm missing something, currently, to update I can assign each attribute separately and do a commit, but this is inconvenient because you cannot assign all of them at once (`record.name = 'foo'; session.commit()`) .
Or, I have to execute session.execute ... but this doesn't seem to make a lot of sense if we already have a record loaded and assigned to a variable. See:

```python
update(User).
...     where(User.name == "sandy").
...     values(fullname = "Sandy Squirrel Extraordinaire")
```

### Databases / Backends / Drivers targeted

all

### Example Use

```python
record = session.query(Model).filter_by(**filters).first()

# proposed syntax
record.update(**attributes) ; session.commit()
# or
session.update(record).values(**attributes)
```

**Comment by zzzeek** [comment]
it's not that hard to make a method that looks like the one you describe but the main issue with it is that insert() and delete() are both idempotent methods, whereas the proposed update(record).values() is not, plus it doesnt seem to imply whether or not the state of "record" actually changes, nor does it make it clear how such a method would interact with the unit of work.

more directly, these use cases are clear:

```py

obj1 = SomeObject()

session.add(obj1)

# if called again, idempotent, no effect
session.add(obj1)


obj2 = session.get(SomeObject, 1),
obj3 = session.get(SomeObject, 1)

# obj2 and ob3 are the same instance

# yet we can call delete() on one or the other, and there is no conflict here,
# it's idempotent
session.delete(obj2)
session.delete(obj3)

# but if code does this, then it's not clear what we're supposed to do
obj4 = session.get(SomeObject, 2)
obj5 = session.get(SomeObject, 3)

# these are conflicting operations, does this raise ?  
obj4.update(foo='bar')
obj5.update(foo='bat')


# What's the value of the "foo" attribute?
obj5.foo ==  ??

```

the SQLAlchemy ORM is based on the notion that the objects you get back from queries are the holders of database state, that is, they are proxies for a row.  So it's most intuitive that to change the value of the row, you mutate that actual object.

**Reply by januszm** [reply]
@zzzeek  thanks for the quick reply and I think your examples are pretty straightforward after all (however I think you meant loading the same record twice with `(SomeObject, 2)`, not 2 and 3 separately):

```
# these are conflicting operations, does this raise ?  
obj4.update(foo='bar')
obj5.update(foo='bat')

# What's the value of the "foo" attribute?
obj5.foo ==  ??
```

`obj5.foo` is `bat`, `obj4.foo` will be `bat` after reload (assuming you wanted to load the same record with different variable names), this is a common problem, an instance of the object will not know about the update until it is reloaded. Last update always wins. An optimistic locking mechanism could be added and built into the framework on top of that.

**Reply by zzzeek** [reply]
oh sorry, I actually wrote the example incorrectly.  the form you proposed was this:

    session.update(obj4).values(foo='bat')

first off, the Session object does not support method chaining, so session.update().values() doesn't make sense in any case.  you can of course run an update statement like this:

    session.execute(update(MyClass).values(...))

but that doesnt automatically grab the primary key of a particular object.      but also it doesnt seem clear why one woudl need session.update() in order to update the atributes on an object - you can just update them directly.  

your original issue was that there is no "update" analogue to "session.add()" / "session.delete()" and it remains that such a method is redundant vs. simply updating the attributes on the target object itself, since as above you still favored session.udpate(object).values() actually changing the state of the object.   Just update the object directly.   

These are patterns that SQLAlchemy didn't invent, they come from other unit-of-work ORMs such as Hibernate, and additionally this is how the ORM has worked for basically 16 years, hundreds of millions of downloads, etc..   So these are not exactly small cosmetic details that are open to just being changed on a whim.     

If you're looking for a more active record type of approach with simple object.save() methods, there's other options out there such as PeeWee and Django ORM.

**Reply by januszm** [reply]
Hmm, I actually used this existing example (from docs) as an inspiration:

```python
session.query(User).filter(User.age == 25)
```

Anyways, my intention was not to propose a specific solution, but rather to start a discussion on facilitating the update process. Like I mentioned above, updating directly on the object only works with one attribute (e.g. `record.foo = 1`) and mass-updates via session.update doesn't work on the loaded record.

> If you're looking for a more active record type of approach with simple object.save() methods, there's other options out there such as PeeWee and Django ORM.

Ah, maybe that answers my questions, yes I am looking for an ActiveRecord equivalent for Python and I started with SQLAlchemy because my applications are not related to Django in any way  (although I am aware that I can use Django ORM independently). It seems that I will probably migrate to a different ORM because SQLAlchemy has a completely different philosophy and assumptions. That probably sums up the discussion here. Thanks!

**Reply by CaselIT** [reply]
> ```
> session.execute(update(MyClass).values(...))
> ```
> 
> 
>     
>   
> 
> but that doesnt automatically grab the primary key of a particular object

not sure if `update` could do that.
Meaning maybe `update(model_intance).value(...)` could use the pk of model_instance to create `update(model_class).where(the_pk).values(...)`. Maybe a different function instead of update `update` would be preferable in that case?
